go-controller/pkg/ovn/address_set.go

```diff
@@ -0,0 +1,234 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+type addressSet struct {
+	sync.Mutex
+
+	name      string
+	hashName  string
+	ports     map[string]net.IP // logical port :: pod IP
+	destroyed bool
+}
+
+// hash the provided input to make it a valid addressSet name.
+func hashedAddressSet(s string) string {
+	return hashForOVN(s)
+}
+
+type addressSetIterFn func(name, namespace, suffix string)
+
+// forEachAddressSetUnhashedName will pass the unhashedName, namespaceName and
+// the first suffix in the name to the 'iteratorFn' for every address_set in
+// OVN. (Each unhashed name for an addressSet can be of the form
+// namespaceName.suffix1.suffix2. .suffixN)
+func forEachAddressSetUnhashedName(iteratorFn addressSetIterFn) error {
+	output, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=external_ids", "find", "address_set")
+	if err != nil {
+		return fmt.Errorf("error reading address sets: "+
+			"stdout: %q, stderr: %q err: %v", output, stderr, err)
+	}
+	for _, addrSet := range strings.Fields(output) {
+		if !strings.HasPrefix(addrSet, "name=") {
+			continue
+		}
+		addrSetName := addrSet[5:]
+		names := strings.Split(addrSetName, ".")
+		addrSetNamespace := names[0]
+		nameSuffix := ""
+		if len(names) >= 2 {
+			nameSuffix = names[1]
+		}
+		iteratorFn(addrSetName, addrSetNamespace, nameSuffix)
+	}
+	return nil
+}
+
+func NewAddressSet(name string, ports map[string]net.IP) (*addressSet, error) {
+	if ports == nil {
+		ports = make(map[string]net.IP)
+	}
+	as := &addressSet{
+		name:     name,
+		hashName: hashedAddressSet(name),
+		ports:    ports,
+	}
+
+	klog.V(5).Infof("New(%s/%s) with %v", as.hashName, as.name, ports)
+
+	addressSet, stderr, err := util.RunOVNNbctl("--data=bare",
+		"--no-heading", "--columns=_uuid", "find", "address_set",
+		"name="+as.hashName)
+	if err != nil {
+		return nil, fmt.Errorf("find failed to get address set %q, stderr: %q (%v)",
+			as.name, stderr, err)
+	}
+
+	// addressSet has already been created in the database and nothing to set.
+	if addressSet != "" {
+		if err := as.setOrClear(); err != nil {
+			return nil, err
+		}
+	} else {
+		// addressSet has not been created yet. Create it.
+		args := []string{
+			"create",
+			"address_set",
+			"name=" + as.hashName,
+			"external-ids:name=" + as.name,
+		}
+		joinedIPs := as.joinIPs()
+		if len(joinedIPs) > 0 {
+			args = append(args, "addresses="+joinedIPs)
+		}
+		_, stderr, err = util.RunOVNNbctl(args...)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create address set %q, stderr: %q (%v)",
+				as.name, stderr, err)
+		}
+	}
+
+	return as, nil
+}
+
+func (as *addressSet) joinIPs() string {
```

Comment:
dont need this if you just use ip addresses in the address  set struct

Reply:
@trozet still need to quote all of them; so either we quote on insert/delete or we quote when we join...

Reply:
you dont need to quote each IP:
```
[root@ovn-control-plane ~]# ovn-nbctl set address_set a112311a "addresses=1.1.1.1,2.2.2.2,3.3.3.3,4.4.4.4"
_uuid               : 95f4ec1a-13f7-4098-ac98-13d97879aa4f
addresses           : ["1.1.1.1", "2.2.2.2", "3.3.3.3", "4.4.4.4"]
external_ids        : {}
name                : a112311a

[root@ovn-control-plane ~]# ovn-nbctl set address_set a112311a addresses=1.1.1.1,2.2.2.2,3.3.3.3,4.4.4.4,5.5.5.5
_uuid               : 95f4ec1a-13f7-4098-ac98-13d97879aa4f
addresses           : ["1.1.1.1", "2.2.2.2", "3.3.3.3", "4.4.4.4", "5.5.5.5"]
external_ids        : {}
name                : a112311a

```

Reply:
you do for IPv6 addresses, for some reason